runtime/src/accounts_partition.rs

```diff
@@ -0,0 +1,718 @@
+        // 5 (40..50)        |  0..0   *2|<20>..<20>    |<19>..<19> *3 <-+- gapped
+        //                   |  0..4]    |<20>..24      |<19>..24      <-+
+        //                   | [4..5]    |  24..25      |  24..25
+        //                   | [5..6]    |  25..26      |  25..26
+        //
+        // NOTE: <..> means the adjusted slots
+        //
+        // *1: The range of parent_bank.slot() and current_bank.slot() is firstly
+        //     split by the epoch boundaries and then the split ones are given to us.
+        //     The original ranges are denoted as [...]
+        // *2: These are marked with generated_for_gapped_epochs = true.
+        // *3: This becomes no-op partition
+        start_partition_index -= 1;
+        if generated_for_gapped_epochs {
+            assert_eq!(start_slot_index, end_slot_index);
+            end_partition_index -= 1;
+        }
+    }
+
+    (start_partition_index, end_partition_index, partition_count)
+}
+
+/// used only by filler accounts in debug path
+/// previous means slot - 1, not parent
+#[cfg(test)]
+pub(crate) fn variable_cycle_partition_from_previous_slot(
+    epoch_schedule: &EpochSchedule,
+    slot: Slot,
+) -> Partition {
+    // similar code to Bank::variable_cycle_partitions
+    let (current_epoch, current_slot_index) = epoch_schedule.get_epoch_and_slot_index(slot);
+    let (parent_epoch, mut parent_slot_index) =
+        epoch_schedule.get_epoch_and_slot_index(slot.saturating_sub(1));
+    let cycle_params = rent_single_epoch_collection_cycle_params(
+        current_epoch,
+        epoch_schedule.get_slots_in_epoch(current_epoch),
+    );
+
+    if parent_epoch < current_epoch {
+        parent_slot_index = 0;
+    }
+
+    let generated_for_gapped_epochs = false;
+    get_partition_from_slot_indexes(
+        cycle_params,
+        parent_slot_index,
+        current_slot_index,
+        generated_for_gapped_epochs,
+    )
+}
+
+/// return all end partition indexes for the given partition
+/// partition could be (0, 1, N). In this case we only return [1]
+///  the single 'end_index' that covers this partition.
+/// partition could be (0, 2, N). In this case, we return [1, 2], which are all
+/// the 'end_index' values contained in that range.
+/// (0, 0, N) returns [0] as a special case.
+/// There is a relationship between
+/// 1. 'pubkey_range_from_partition'
+/// 2. 'partition_from_pubkey'
+/// 3. this function
+pub(crate) fn get_partition_end_indexes(partition: &Partition) -> Vec<PartitionIndex> {
+    if partition.0 == partition.1 && partition.0 == 0 {
+        // special case for start=end=0. ie. (0, 0, N). This returns [0]
+        vec![0]
+    } else {
+        // normal case of (start, end, N)
+        // so, we want [start+1, start+2, ..=end]
+        // if start == end, then return []
+        (partition.0..partition.1).map(|index| index + 1).collect()
+    }
+}
+
+pub(super) fn rent_single_epoch_collection_cycle_params(
```

Comment:
super is just the crate now, so let's update this to be less confusing. similarly for the other `pub(super)`s

Reply:
Whoops. Relic left over from when I tried to structure this as a sub-module of bank. Fixed